cmds-42/usr.etc/nmserver/man/nmtime.1
.TH NMTIME 1 2/22/88
.CM 4
.SH NAME
nmtime \- Timing program for the Mach network server.
.SH SYNOPSIS
.B nmtime
[ -e ]
.SH DESCRIPTION
.I nmtime
can be used to test and obtain timings for a
.I network server.
To operate, one instance of
.I nmtime
should be put in passive echo mode by using the
.B -c
command line switch or the 
.B E
command from the
.I nmtime
prompt, and the other instance should be used interactively 
to send messages.

.I nmtime
is interactive; all commands can be listed by entering 
.I ?
or
.I h
at the prompt.
Commands can be in upper or lower case; several commands
can be typed on the same line, provided that spaces are used to 
disambiguate the end of string arguments.
.SH HISTORY
.TP 
22-Feb-88  Daniel Julin (dpj) at Carnegie-Mellon University
Created.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.etc/nmserver/utils/nmtime.c
/* 
 * Mach Operating System
 * Copyright (c) 1987 Carnegie-Mellon University
 * All rights reserved.  The CMU software License Agreement specifies
 * the terms and conditions for use and redistribution.
 */
/*
 * nmtime.c
 *
 * $Source: /os/osdev/LIBS/libs-7/etc/nmserver/utils/RCS/nmtime.c,v $
 *
 */
#ifndef	lint
static char     rcsid[] = "$Header: nmtime.c,v 1.1 88/09/30 15:47:11 osdev Exp $";
#endif not lint
/*
 * Program to measure the timing of various network server operations.
 */

/*
 * HISTORY: 
 * 21-Jul-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Renamed variable to "inline" to "is_inline" to avoid conflicts
 *	with compilers that think "inline" is a keyword.
 *
 *  8-Dec-87  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Made faster by reusing as much data as possible in the
 *	message buffer.
 *
 *  7-Dec-87  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Created.
 *
 */


#include	"debug.h"
#include	"trace.h"
#include	"nm_defs.h"

#include	<stdio.h>
#include	<mach.h>
#include	<servers/netname.h>
#include	<sys/message.h>
#include	<msg_type.h>
#include	<sys/time.h>

#define readln(p) while (((getc(p))!='\n') && (!feof(p)))
#define NETNAME_NAME	"NEW_NETWORK_NAME_SERVER"


/*
 * ID's for test messages.
 */
#define	NM_TIME_READ	10001
#define	NM_TIME_WRITE	10002


/*
 * Global variables
 */
static port_t		nn_port;		/* port to new network server */
static char		service_name[80];	/* nmtime service */
static char		echo_name[80];		/* name of the echo server */
static port_t		echo_port;		/* port for current echo server */
static port_t		reply_port;
static struct {
	boolean_t	dealloc;		/* dealloc ool sections */
} options;


/*
 * Message buffer.
 */
#define	TIME_DATA_MAX	\
		(MSG_SIZE_MAX - (sizeof(msg_header_t) + sizeof(msg_type_long_t)))

struct msg_buff {
	msg_header_t	header;
	msg_type_long_t	desc;
	union {
		int		size;
		char		inline_data[TIME_DATA_MAX];
		char		*ool_ptr;
	}		msg_data;
} msg_buff;

#define DEALLOC_OOL(buf) {						\
	(void) vm_deallocate(task_self(),buf.msg_data.ool_ptr,		\
				buf.desc.msg_type_long_number);		\
}


/*
 * Tracing values.
 */
int	trace_recursion_level = 0;
#undef	tracing_on
int	tracing_on = 0;


/*
 * input --
 *
 * Obtain user input, allowing multiple inputs on the same line.
 *
 * Parameters:
 *
 * prompt: string to use as a prompt.
 * format: specification for data required.
 * OUT var: address where to put the data read.
 *
 * Results:
 *
 * Side effects:
 *
 * May print a prompt on stdout, and start a read on stdin.
 *
 * Design:
 *
 * Note:
 *
 */
void input(prompt, format, var)
char *prompt;
char *format;
char *var;
BEGIN("input")
	char c;
	int fcnt;

	while ((stdin->_cnt > 0) &&
	    ((((c = *(stdin)->_ptr&0377) == ' ') || (c == '\n'))))
		getc(stdin);
	fcnt = stdin->_cnt;

	if (fcnt == 0) {
		fprintf(stdout, "%s", prompt);
		fflush(stdout);
	}
	fscanf(stdin, format, var);

	RET;
END


/*
 * printmenu --
 *
 * Parameters: none
 *
 * Results:
 *
 * Side effects:
 *
 * Print a command menu on stdout.
 *
 * Design:
 *
 * Note:
 *
 */
void printmenu()
BEGIN("printmenu")
	fprintf(stdout, "\n");
	fprintf(stdout, "This programs allows to measure the performance of a network server.\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "Available commands:\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    H,?.......Print this message\n");
	fprintf(stdout, "    C.........(Re)connect to network server\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    L.........Look up echo server <host>\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Options:\n");
	fprintf(stdout, "       OD.........Deallocate OOL sections received <T/F>\n");
	fprintf(stdout, "       OE.........Set service name for echo server <name>\n");
	fprintf(stdout, "       OL.........List option settings\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Read transactions (data in the reply):\n");
	fprintf(stdout, "       RI.........IPC read <size> <count>\n");
	fprintf(stdout, "       RR.........RPC read <size> <count>\n");
	fprintf(stdout, "\n");
	fprintf(stdout, " Write transactions (data in the request):\n");
	fprintf(stdout, "       WI.........IPC write <size> <count>\n");
	fprintf(stdout, "       WR.........RPC write <size> <count>\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    E.........Start echo service (does not return)\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "    Q.........Quit\n");
	fprintf(stdout, "\n");
	RET;
END


/*
 * nmconnnect --
 *
 * Connect to the network name server.
 */
void nmconnect()
BEGIN("nmconnect")
	kern_return_t	kr;

	kr = netname_look_up(name_server_port, "", NETNAME_NAME, &nn_port);
	if (kr == NETNAME_SUCCESS) {
		fprintf(stdout,
			"Look up of new network server succeeds, port = %x.\n",
			nn_port);
	} else {
		fprintf(stdout,
	"Look up of new network server fails, using old network server, port = %x.\n",
			name_server_port);
		nn_port = name_server_port;
	}
	RET;
END


/*
 * lookupecho --
 *
 * Find a new echo server.
 *
 * Parameters:
 *
 * name: name of machine to look for, or NULL.
 *
 * Results:
 *
 * Side effects:
 *
 * May prompt for the name of the machine to control. Sets echo_port and 
 * echo_name to the appropriate value.
 *
 * Design:
 *
 * Note:
 *
 */
void lookupecho()
BEGIN("lookupecho")
	kern_return_t		kr;

	input("Echo server host: ", "%79s", echo_name);

	fprintf(stdout,"Looking up \"%s\" at %s ...\n",service_name,echo_name);
	kr = netname_look_up(nn_port,echo_name,service_name,&echo_port);
	if (kr == NETNAME_SUCCESS) {
		fprintf(stdout,"    [Successful]\n");
	} else {
		fprintf(stdout,"    [Can't find it!  kr = %1d]\n",kr);
		readln(stdin);
	}
	fflush(stdout);

	RET;
END


/*
 * genericopts --
 *
 * Execute any option setting or listing operation.
 */
void genericopts()
BEGIN("genericopts")
	char		subcmd[2];
	boolean_t	*subopt;
	char		newval[2];

#define	OPTVAL(opt)	(options.opt ? "TRUE" : "FALSE")

	subopt = NULL;

	input("Option subcommand: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'D':
		case 'd':
			subopt = &options.dealloc;
			break;
		case 'E':
		case 'e':
			input("Echo service name: ", "%79s", service_name);
			break;
		case 'L':
		case 'l':
			fprintf(stdout,"  Deallocate OOL sections received: %s\n",
								OPTVAL(dealloc));
			fprintf(stdout,"  Echo service name: %s\n",service_name);
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			RET;
	}

	if (subopt != NULL) {
		input("Option value (T/F): ", "%1s", newval);
		switch(newval[0]) {
			case 'T':
			case 't':
				*subopt = TRUE;
				break;
			case 'F':
			case 'f':
				*subopt = FALSE;
				break;
			default:
				fprintf(stdout, "Invalid option value.\n");
				readln(stdin);
				RET;
		}
	}

	RET;
END


/*
 * genericread --
 *
 * Execute any read transaction.
 */
void genericread()
BEGIN("genericread")
	char		subcmd[2];
	int		input_size;
	int		count;
	int		msg_type;
	int		msg_size;
	int		data_size;
	kern_return_t	kr;
	int		iteration;
	struct timeval	start_time, end_time;
	int		total_time;

	if (echo_port == PORT_NULL) {
		fprintf(stdout, "Must first look up an echo server\n");
		readln(stdin);
		RET;
	}

	input("Msg type: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'I':
		case 'i':
			msg_type = MSG_TYPE_NORMAL;
			break;
		case 'R':
		case 'r':
			msg_type = MSG_TYPE_RPC;
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			RET;
	}

	input("Data size: ", "%d", &input_size);
	if (input_size <= 0) {
		fprintf(stdout,"Invalid size.\n");
		readln(stdin);
		RET;
	}

	data_size = input_size;
	if (data_size > TIME_DATA_MAX) {
		fprintf(stdout, "Size too large for an inline message.\n");
		readln(stdin);
		RET;
	}

	input("Repeat count: ", "%d", &count);
	if (count <= 0) {
		fprintf(stdout,"Invalid count.\n");
		readln(stdin);
		RET;
	}
	
	/*
	 * Set up the fields in the message that never change.
	 * The echo server must preserve those fields.
	 */
	msg_buff.header.msg_simple = TRUE;
	msg_buff.header.msg_local_port = reply_port;
	msg_buff.header.msg_remote_port = echo_port;
	msg_buff.header.msg_id = NM_TIME_READ;

	msg_buff.desc.msg_type_long_name = MSG_TYPE_BYTE;
	msg_buff.desc.msg_type_long_size = 8;
	msg_buff.desc.msg_type_header.msg_type_inline = TRUE;
	msg_buff.desc.msg_type_header.msg_type_longform = TRUE;
	msg_buff.desc.msg_type_header.msg_type_deallocate = FALSE;

	msg_size = sizeof(msg_header_t) + sizeof(msg_type_long_t) + 4;

	(void) gettimeofday(&start_time, 0);

	for (iteration = 1; iteration <= count; iteration++) {
		msg_buff.header.msg_size = msg_size;
		msg_buff.header.msg_type = msg_type;
		msg_buff.desc.msg_type_long_number = 4;
		msg_buff.msg_data.size = htonl(data_size);

		kr = msg_rpc(&msg_buff, MSG_OPTION_NONE, MSG_SIZE_MAX, 0, 0);
		if (kr != SEND_SUCCESS) {
			fprintf(stdout,"msg_rpc fails on message %d, kr=%d.\n", 
								iteration, kr);
			readln(stdin);
			break;
		}
	}

	(void)gettimeofday(&end_time, 0);
	total_time = ((end_time.tv_sec - start_time.tv_sec) * 1000) +
				((end_time.tv_usec - start_time.tv_usec) / 1000);
	fprintf(stdout,"Sending %d messages took %d milliseconds.\n",
							count, total_time);

	RET;
END


/*
 * genericwrite --
 *
 * Execute any write transaction.
 */
void genericwrite()
BEGIN("genericwrite")
	char		subcmd[2];
	int		input_size;
	int		count;
	int		msg_type;
	int		msg_size;
	boolean_t	is_inline;
	int		data_size;
	char		*ool_ptr = NULL;
	kern_return_t	kr;
	int		iteration;
	struct timeval	start_time, end_time;
	int		total_time;

	if (echo_port == PORT_NULL) {
		fprintf(stdout, "Must first look up an echo server\n");
		readln(stdin);
		RET;
	}

	input("Msg type: ", "%1s", subcmd);
	switch(subcmd[0]) {
		case 'I':
		case 'i':
			msg_type = MSG_TYPE_NORMAL;
			break;
		case 'R':
		case 'r':
			msg_type = MSG_TYPE_RPC;
			break;
		default:
			fprintf(stdout, "Invalid subcommand. Type ? for help.\n");
			readln(stdin);
			RET;
	}

	input("Data size (negative size = OOL): ", "%d", &input_size);
	if (input_size < 0) {
		is_inline = FALSE;
		data_size = - input_size;
	} else {
		is_inline = TRUE;
		data_size = input_size;
		if (data_size > TIME_DATA_MAX) {
			fprintf(stdout, "Size too large for an inline message.\n");
			readln(stdin);
			RET;
		}
	}

	input("Repeat count: ", "%d", &count);
	if (count <= 0) {
		fprintf(stdout,"Invalid count.\n");
		readln(stdin);
		RET;
	}
	
	/*
	 * Set up the fields in the message that never change.
	 * The echo server must preserve those fields.
	 */
	msg_buff.header.msg_local_port = reply_port;
	msg_buff.header.msg_remote_port = echo_port;
	msg_buff.header.msg_id = NM_TIME_WRITE;

	msg_buff.desc.msg_type_long_name = MSG_TYPE_BYTE;
	msg_buff.desc.msg_type_long_size = 8;
	msg_buff.desc.msg_type_long_number = data_size;
	msg_buff.desc.msg_type_header.msg_type_longform = TRUE;
	msg_buff.desc.msg_type_header.msg_type_deallocate = FALSE;

	if (data_size == 0) {
		msg_size = sizeof(msg_header_t);
	} else {
		if (is_inline) {
			msg_size = sizeof(msg_header_t) + 
					sizeof(msg_type_long_t) + data_size;
			msg_buff.desc.msg_type_header.msg_type_inline = TRUE;
		} else {
			ool_ptr = NULL;
			kr = vm_allocate(task_self(),&ool_ptr,data_size,TRUE);
			if (kr != KERN_SUCCESS) {
				fprintf(stdout,"vm_allocate failed: kr=%d\n", kr);
			readln(stdin);
				RET;
			}
			msg_size = sizeof(msg_header_t) +
					sizeof(msg_type_long_t) + sizeof(char *);
			msg_buff.desc.msg_type_header.msg_type_inline = FALSE;
			msg_buff.msg_data.ool_ptr = ool_ptr;
		}
	}

	(void) gettimeofday(&start_time, 0);

	for (iteration = 1; iteration <= count; iteration++) {
		msg_buff.header.msg_simple = is_inline;
		msg_buff.header.msg_size = msg_size;
		msg_buff.header.msg_type = msg_type;

		kr = msg_rpc(&msg_buff, MSG_OPTION_NONE, sizeof(msg_header_t), 0, 0);
		if (kr != SEND_SUCCESS) {
			fprintf(stdout,"msg_rpc fails on message %d, kr=%d.\n", 
								iteration, kr);
			readln(stdin);
			break;
		}
	}

	(void)gettimeofday(&end_time, 0);
	total_time = ((end_time.tv_sec - start_time.tv_sec) * 1000) +
				((end_time.tv_usec - start_time.tv_usec) / 1000);
	fprintf(stdout,"Sending %d messages took %d milliseconds.\n",
							count, total_time);

	if (ool_ptr != NULL) {
		kr = vm_deallocate(task_self(),ool_ptr,data_size);
		if (kr != KERN_SUCCESS) {
			fprintf(stdout,"vm_deallocate failed: kr=%d\n", kr);
			readln(stdin);
		}
	}

	RET;
END


/*
 * echostart --
 *
 * Start echoing messages for another instance of the timer program.
 *
 * Note:
 *	This procedure never returns.
 */
void echostart()
BEGIN("echostart")
	kern_return_t		kr;
	port_t			server_port;
	port_t			dummy_port;
	int			size;

	kr = port_allocate(task_self(), &server_port);
	if (kr != KERN_SUCCESS) {
		fprintf(stdout,"port_allocate of server port failed, kr=%d.\n",kr);
		exit(1);
	}
	kr = port_disable(task_self(), server_port);
	if (kr != KERN_SUCCESS) {
		fprintf(stdout,"port_disable of server port failed, kr=%d.\n",kr);
		exit(1);
	}

	kr = netname_check_in(nn_port, service_name, 0, server_port);
	if (kr != KERN_SUCCESS) {
		fprintf(stdout,"checkin of server port failed, kr=%d.\n",kr);
		exit(1);
	}

	kr = port_allocate(task_self(), &dummy_port);
	kr = port_disable(task_self(), dummy_port);

	msg_buff.header.msg_size = sizeof(msg_header_t);
	msg_buff.header.msg_local_port = server_port;
	msg_buff.header.msg_remote_port = dummy_port;

	fprintf(stdout,"Starting echo service - hit ^C to abort.\n");
	fflush(stdout);

	while (1) {
		msg_buff.header.msg_simple = TRUE;
		msg_buff.header.msg_type = MSG_TYPE_NORMAL;
		kr = msg_rpc(&msg_buff, MSG_OPTION_NONE, MSG_SIZE_MAX, 0, 0);
		if (kr != SEND_SUCCESS) {
			fprintf(stdout,"msg_rpc failed, kr=%d.\n",kr); 
			RET;
		}

		switch (msg_buff.header.msg_id) {
			case NM_TIME_READ:
				/*
				 * We can reuse the msg_type from the request.
				 * Just fix the sizes.
				 */
				size = ntohl(msg_buff.msg_data.size);
				msg_buff.header.msg_size = sizeof(msg_header_t) + 
						sizeof(msg_type_long_t) + size;
				msg_buff.desc.msg_type_long_number = size;
				break;

			case NM_TIME_WRITE:
				if (msg_buff.header.msg_size !=
							sizeof(msg_header_t)) {
					/*
					 * We need to reset the data portion
					 * of the msg before sending it back.
					 */
					if ((options.dealloc) && 
					(msg_buff.header.msg_simple == FALSE)) {
						DEALLOC_OOL(msg_buff);
					}
					msg_buff.header.msg_size = 
							sizeof(msg_header_t);
				}
				break;

			default:
				fprintf(stdout," Received a message with unknown id: %d.\n", msg_buff.header.msg_id);
				fflush(stdout);
				break;
		}
	}

END


/*
 * initialize --
 *
 * Initialize all the test functions.
 */
int initialize()
BEGIN("initialize")
	kern_return_t		kr;

	/*
	 * Connect to name server.
	 */
#ifndef	NeXT
	init_netname(PORT_NULL);
#endif	NeXT
	nmconnect();

	/*
	 * Global variables.
	 */
	echo_port = PORT_NULL;
	kr = port_allocate(task_self(),&reply_port);
	if (kr != KERN_SUCCESS) {
		fprintf(stdout,"port_allocate of reply port failed, kr=%d\n",kr);
		exit(1);
	}
	kr = port_disable(task_self(),reply_port);
	if (kr != KERN_SUCCESS) {
		fprintf(stdout,"port_disable of reply port failed, kr=%d\n",kr);
		exit(1);
	}
	options.dealloc = FALSE;
	strcpy(service_name,"NM_TIME");

	RETURN(1);
END


/*
 * main --
 *
 * Main loop.
 *
 * Parameters:
 *
 * Results:
 *
 * Side effects:
 *
 * Design:
 *
 * Note:
 *
 */
main(argc, argv)
int argc;
char **argv;
BEGIN("main")
	int quit;
	char cmd[2];

	(void)initialize();

	if (argc > 1) {
		if ((argc == 2) && (strcmp(argv[1],"-e") == 0)) {
			echostart();
		} else {
			fprintf(stdout,"Usage: %s [-e]\n",argv[0]);
			exit(1);
		}
	}

	quit = FALSE;
	do {
		input("> ", "%1s", cmd);
		switch (cmd[0]) {
		case 'H':
		case 'h':
		case '?':
			printmenu();
			break;
		case 'C':
		case 'c':
			nmconnect();
			break;
		case 'L':
		case 'l':
			lookupecho();
			break;
		case 'O':
		case 'o':
			genericopts();
			break;
		case 'R':
		case 'r':
			genericread();
			break;
		case 'W':
		case 'w':
			genericwrite();
			break;
		case 'E':
		case 'e':
			echostart();
			break;
		case 'Q':
		case 'q':
			quit = TRUE;
			break;
		default:
			fprintf(stdout, "Invalid command. Type ? for list.\n");
			readln(stdin);
			break;
		}
	} 
	while (!quit);

	fprintf(stdout, "Terminated.\n");
	exit(0);
END


